### PR TITLE
Change IPointer interface method to return nint

### PR DESCRIPTION
**Pull request by lonitra**
Fixes https://github.com/dotnet/winforms/issues/11983

This issue occurs because System.Private.Windows.Core is included in the ref as C# needs the types to be there when it compiles interface casts (https://github.com/dotnet/winforms/pull/10673), but this also causes issues for us in CLI/CLR C++ mixed mode as shown in issue. 

This is a workaround that updates our `IPointer<T>` interface to avoid returning a generic as it present compile issues in CLI/CLR C++ mixed mode. This looks to be a limitation that is in mix mode as this error occurs with non WindowsForms related classes that have similar implementation. This workaround mitigates this issue while coordination is happening to determine a fix.

Tested by building dlls from change and replacing System.Drawing.Common.dll and System.Private.Windows.Core.dll in C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\9.0.0-rc.2.24428.1\ref\net9.0
### Before

image

### After

image

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/winforms/pull/12103)

### CODE REVIEW

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs [hunk 1]
```diff
@@ -4,14 +4,21 @@
...
 /// </remarks>
-internal unsafe interface IPointer<TPointer> where TPointer : unmanaged
+internal unsafe interface IPointer<TPointer> : IPointer where TPointer : unmanaged
+{
+}
+
+/// <summary>
+///  Used to indicate ownership of a native resource pointer.
+/// </summary>
+internal interface IPointer
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs:21]: Why is this necessary?

File: src/System.Private.Windows.Core/src/Windows/Win32/Graphics/Gdi/RegionScope.cs [hunk 2]
```diff
@@ -62,7 +62,7 @@ public RegionScope(HDC hdc)
     /// </summary>
     public RegionScope(IPointer<GpRegion> region, IPointer<GpGraphics> graphics)
     {
-        InitializeFromGdiPlus(region.Pointer, graphics.Pointer);
+        InitializeFromGdiPlus((GpRegion*)region.Pointer, (GpGraphics*)graphics.Pointer);
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Graphics/Gdi/RegionScope.cs:65]: Don't want to do hard casts here. Should use the extension method to ensure type safety.

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Windows.Win32.Foundation;
+internal static unsafe class IPointerExtensions
+{
+    public static T* GetPointer<T>(this IPointer pointer) where T : unmanaged => (T*)pointer.Pointer;
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointerExtensions.cs:7]: Presumably this can be `IPointer<T>` to help ensure we're using the right types.

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Windows.Win32.Foundation;
+internal static unsafe class IPointerExtensions
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointerExtensions.cs:5]: Should probably be `PointerExtensions`.

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs (hunk 1 above)
Comment (lonitra) [reply src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs:21]: 🤦‍♀️ this isn't needed. I missed that the extension method can have IPointer<T>

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs [hunk 5]
```diff
@@ -13,5 +13,6 @@ namespace Windows.Win32.Foundation;
 /// </remarks>
 internal unsafe interface IPointer<TPointer> where TPointer : unmanaged
 {
-    TPointer* Pointer { get; }
+    [Obsolete("Use extension method GetPointer on IPointer<T> to get the typed pointer instead.")]
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs:16]: We still need a comment describing why we're going through these contortions

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
Just missing a comment describing the scenario we're working around

File: src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs [hunk 6]
```diff
@@ -13,5 +13,10 @@ namespace Windows.Win32.Foundation;
 /// </remarks>
 internal unsafe interface IPointer<TPointer> where TPointer : unmanaged
 {
-    TPointer* Pointer { get; }
+    // This interface method must return a nint instead of a typed pointer directly because
+    // CLI/CLR C++ mixed code cannot compile generic pointers and
```
Comment (JeremyKuhne) [inline src/System.Private.Windows.Core/src/Windows/Win32/Foundation/IPointer.cs:17]:
```suggestion
    // C++/CLI cannot compile when it encounters generic pointers, even in internal interfaces.
```

**Review by JeremyKuhne — APPROVED** [review]
